### PR TITLE
Add GC_gcj_vector_malloc to support marking complex (value types with…

### PR DESCRIPTION
**Pull request by joncham**
… reference fields) arrays.

Used by Mono PR: https://github.com/Unity-Technologies/mono/pull/1137

### CODE REVIEW

**Comment by joncham** [comment]
@ivmai is this the best way to expose a GCJ like allocator that is marked via a custom mark proc?

**Comment by joncham** [comment]
My mark proc is a work in progress here: https://github.com/Unity-Technologies/mono/pull/1137/files#diff-d5708d26cfbc2037ac98c34b2e708387R120

File: include/gc_gcj.h [hunk 1]
```diff
@@ -89,12 +91,18 @@ GC_API GC_ATTR_MALLOC GC_ATTR_ALLOC_SIZE(1) void * GC_CALL
...
+GC_gcj_vector_malloc (size_t lb,
+	void * ptr_to_struct_containing_descr);
+
 /* The kind numbers of normal and debug gcj objects.            */
 /* Useful only for debug support, we hope.                      */
 GC_API int GC_gcj_kind;
 
 GC_API int GC_gcj_debug_kind;
 
+GC_API int GC_gcj_vector_kind;
```
Comment (ivmai) [inline include/gc_gcj.h]: It's better to export setter and getter instead.

File: gcj_mlc.c [hunk 2]
```diff
@@ -121,6 +132,38 @@ GC_API void GC_CALL GC_init_gcj_malloc(int mp_index,
     UNLOCK();
 }
 
+/* Caller does not hold allocation lock. */
+GC_API void GC_CALL GC_init_gcj_vector (int mp_index,
```
Comment (joncham) [inline gcj_mlc.c]: I don't think this is worth upstreaming, but in @ivmai in general should I put something custom like this in bdwgc itself? Or should I implement in my client code? If in bdwgc should I move to a separate file?

**Comment by ivmai** [comment]
> I don't think this is worth upstreaming, but in @ivmai in general should I put something custom like this in bdwgc itself? Or should I implement in my client code? If in bdwgc should I move to a separate file?

Right, this requires code refactoring (e.g. code deduplication) if ever considered for upstream.
It is better not to put custom login in bdwgc if it could be placed to the client code. It is not always attainable.

PS. Periodically I look at downstreams for patches that worth upstreaming.

**Pull request merged**